vllm/model_executor/layers/fused_moe.py

```diff
@@ -0,0 +1,340 @@
+    - num_experts: The total number of experts.
+
+    Returns:
+    - sorted_token_ids: A tensor containing the sorted token indices according to their allocated expert.
+    - expert_ids: A tensor indicating the assigned expert index for each token.
+    - num_tokens_post_padded: The total number of tokens after padding, ensuring divisibility by block_size.
+
+    This function pads the number of tokens that each expert needs to process so that it is divisible by block_size. 
+    Padding ensures that during block matrix multiplication, the dimensions align correctly.
+
+    Example:
+    Given topk_ids = [[2, 3, 4], [1, 2, 4], [1, 3, 4], [1, 2, 3]], block_size = 4, and num_experts = 4:
+    - We initially have 12 tokens (after repeating 'top_k' times) and 4 experts, with each expert needing to process 3 tokens.
+    - As block_size is 4, we pad 1 token for each expert.
+    - First, flatten topk_ids to [2, 3, 4, 1, 2, 4, 1, 3, 4, 1, 2, 3].
+    - Then append padding tokens [1, 2, 3, 4] at the end, resulting in [2, 3, 4, 1, 2, 4, 1, 3, 4, 1, 2, 3 | 1, 2, 3, 4].
+    - After sorting by expert index, we obtain token_ids [3, 6, 9, 12, 0, 4, 10, 13, 1, 7, 11, 14, 2, 5, 8, 15]. 
+        Tokens 12-14 are non-existent (padding) and are ignored in the subsequent matrix multiplication.
+    - The padding ensures that the total number of tokens is now divisible by block_size for proper block matrix operations.
+    """
+    cnts = torch.zeros(topk_ids.shape[0],
+                       num_experts,
+                       dtype=topk_ids.dtype,
+                       device=topk_ids.device)
+    cnts.scatter_(1, topk_ids, 1)
+    tokens_per_expert = cnts.sum(dim=0)
+    tokens_per_expert_post_alig = torch.floor_divide(
+        tokens_per_expert + block_size - 1, block_size) * block_size
+
+    cumsum = tokens_per_expert_post_alig.cumsum(0)
+    num_tokens_post_padded = cumsum[-1]
+    max_tokens_post_padded = (topk_ids.numel() + num_experts *
+                              (block_size - 1))
+    expert_ids = torch.zeros(max_tokens_post_padded,
+                             dtype=topk_ids.dtype,
+                             device=topk_ids.device)
+    # expert_ids[cumsum[:-1]] = 1
+    ones = torch.ones_like(expert_ids)
+    expert_ids.scatter_add_(0, cumsum[:-1], ones)
+    expert_ids = expert_ids.cumsum(0)
+
+    cumsum = (tokens_per_expert_post_alig - tokens_per_expert).cumsum(0)
+
+    padded_tokens = torch.zeros(max_tokens_post_padded - topk_ids.numel(),
+                                dtype=topk_ids.dtype,
+                                device=topk_ids.device)
+    ones = torch.ones_like(padded_tokens)
+    padded_tokens.scatter_add_(0, cumsum[:-1], ones)
+    padded_tokens = padded_tokens.cumsum(0)
+
+    sorted_token_ids = torch.cat([topk_ids.view(-1), padded_tokens]).argsort()
+
+    return sorted_token_ids, expert_ids, num_tokens_post_padded
+
+
+cache1 = None
+cache2 = None
+cache3 = None
```

Comment:
A quick question: What are these caches for? Does pre-allocating them improve performance?

Reply:
no performance improvement. i find it may reduces memory fragmentation and the likelihood of out-of-memory issues with large batches

Reply:
i have remove them to avoid issues  that may be caused by global variables.